### DISCUSSION TITLE
Are there any restrictions on the conditions for Match()?

### DISCUSSION THREAD

**Discussion by trobinpl** [NRules/NRules#339] [Q&A] [answered]
I wanted to set up a simple notifications system inside my app using NRules. The backbone of the whole thing were supposed to be rules checking if user has a given type of notifications enabled. Due to multiple notifications channels this information is stored in fact inside a dictionary. In my rules I hade something like this
`Match(() => user, u => u.NotificationsSettings["EventType"].SomeChannel)` where `SomeChannel` is a bool value. When I added all my facts and fired rules the ones I was suspecting to fire, didn't.

When I was testing I added a simple bool value "directly" to the fact and changed Match to something like `Match(() => user, u => u.TestBool)`. With this approach I've got exactly the result I was expecting.

I feel like I'm missing some fundamental concept about the rules, matching and firing? Or maybe there's some limitation on how you should write the matchers due to how the rules are stored internally? Did I missed docs about this?

**Comment by snikolayev** [comment]
@trobinpl no, there are no restrictions. I created this simple scenario that demonstrates. It also prints diagnostics when rule conditions are evaluated. You can play with it to see what the problem with your rule or your user fact is.

```c#
using NRules;
using NRules.Fluent;
using NRules.Fluent.Dsl;

var repository = new RuleRepository();
repository.Load(x => x.From(typeof(TestRule).Assembly));

var factory = repository.Compile();
var session = factory.CreateSession();

session.Events.LhsExpressionEvaluatedEvent += (sender, args) =>
{
    Console.WriteLine($"LHS expression evaluated: {args.Expression}, result: {args.Result}");
};

var user = new User{Name = "John Do"};
user.NotificationsSettings.Add("EventType", new EventInfo{SomeChannel = true});

session.Insert(user);
session.Fire();

public class EventInfo
{
    public bool SomeChannel { get; set; }
}


public class User
{
    public string Name { get; set; }
    public Dictionary<string, EventInfo> NotificationsSettings { get; set; } = new Dictionary<string, EventInfo>();
}

public class TestRule : Rule
{
    public override void Define()
    {
        var user = default(User);

        When()
            .Match(() => user, u => u.NotificationsSettings["EventType"].SomeChannel);

        Then()
            .Do(ctx => Console.WriteLine($"Matched user {user.Name}"));   
    }
}
```

Output:
```
LHS expression evaluated: user => user.NotificationsSettings.get_Item("EventType").SomeChannel, result: True
Matched user John Do
```

**Comment by trobinpl** [comment]
I played around with the example you've provided and it looks like the issue lays not in the matching condition, but rather in having multiple rules.

I've got correct result when I registered only one rule, however having two almost identical rules loaded

`When().Match(() => user, u => u.NotificationSettings["EventType"].SomeChannel); //SomeChannelRule.cs`
and
`When().Match(() => user, u => u.NotificationSettings["EventType"].OtherChannel); //OtherChannelRule.cs`

I get

```
Notify user1 via Some Channel
Notify user2 via Some Channel
Notify user1 via Other Channel
Notify user2 via Other Channel
```
even when `user1` has the notification settings for `OtherChannel` set to `false`.

What's also weird (at least with my limited understaning) is the diagnostics info

```
LHS expression evaluated: User => User.NotificationSettings.get_Item("EventType").SomeChannel, result: False
LHS expression evaluated: User => User.NotificationSettings.get_Item("EventType").SomeChannel, result: True
LHS expression evaluated: User => User.NotificationSettings.get_Item("EventType").SomeChannel, result: True
LHS expression evaluated: User => User.NotificationSettings.get_Item("EventType").SomeChannel, result: False
```

Shouldn't there be also info about evaluating expressions for the `OtherChannel`?

I've confirmed that both my rules are loaded into the repository. Both of them are `public class`, inherit from `Rule` and override `Define`. I've also tried to tag them with something like `Tag("SomeChannel")` and load rules into multiple rulesets based on that tag, but it also produced the same outcome.

**Reply by snikolayev** [reply]
@trobinpl can you please modify the sample I provided with the changes you outlined, so that it reproduces the issue, and post it here? Would be easier this way instead of me recreating the scenario from your description

**Reply by trobinpl** [reply]
Of course - please take a look at this example repo I've setup to see exactly what is my current approach https://github.com/trobinpl/nrules-issue-repo

When I run the app output is 

```
LHS expression evaluated: User => User.NotificationSettings.get_Item("NewsPosted").InApp, result: True
LHS expression evaluated: User => User.NotificationSettings.get_Item("NewsPosted").InApp, result: True
LHS expression evaluated: User => User.NotificationSettings.get_Item("NewsPosted").InApp, result: False
LHS expression evaluated: User => User.NotificationSettings.get_Item("NewsPosted").InApp, result: False
Notify bothOn via InApp
Notify onlyInApp via InApp
Notify bothOn via Slack
Notify onlyInApp via Slack
```

Sorry for making it a little confusing with my previous explanation! 🙂

**Reply by snikolayev** [reply, marked as answer]
@trobinpl thanks for that repro case - you found a bug! See #340
While I'm fixing it, you can work around, by wrapping those properties into methods and use those in the rules.

```c#
public class NotificationSettings
{
    public bool Slack { get; set; }
    public bool InApp { get; set; }
    public bool HasSlack() => Slack;
    public bool HasApp() => InApp;
}
```

**Reply by trobinpl** [reply]
Och my! That's my first!

I'm glad the repro repo helped! I'll use the workaround for now, waiting for the new release 🙂 

Thanks for developing this amazing project! ❤️
